### PR TITLE
Updates Madras Env to support multiple single agent trainings, parallely 

### PR DESCRIPTION
**Pull request by buridiaditya**
What this PR adds to the Env
1. Updates Madras Env with the auto launch of torcs on a random vacant port.
2. Supports parallel agents.

Why this PR is required is 
1. For parallel training of single agents.

Testing:
1. Clone the baselines and MADRaS repos.
2. Install editable copies of both using `pip install -e .`.
3. run `python -m baselines.run --alg='ddpg' --env='Madras-v0'` from baselines folder for single agent training.
4. run `mpirun -np 4 python -m baselines.run --alg='ddpg' --env='Madras-v0'` from baselines folder for parallel single agent training.

### CODE REVIEW

File: MADRaS/envs/gym_madras.py [hunk 1]
```diff
@@ -75,9 +115,7 @@ def reset(self, prev_step_info=None):
 
         else:
             try:
-                self.ob, self.client =\
-                    TorcsEnv.reset(self, client=self.client, relaunch=True)
-
+		self.ob, self.client = TorcsEnv.reset(self, client=self.client, relaunch=True)
```
Comment (rudrasohan) [inline MADRaS/envs/gym_madras.py]: indentation is off here.

**Review by rudrasohan — COMMENTED** [review]
I tested with baselines. There were no errors but the agent restarts instantly. And at the time of restarting the reset waits for 100 sec.

**Comment by rudrasohan** [comment]
The autostart function is also not working properly.

**Comment by buridiaditya** [comment]
- [BUG] `pkill torcs` kill all torcs instances.
- [BUG] update autostart.sh for auto navigation.
- [BUG] assign window names to torcs.

**Comment by rudrasohan** [comment]
@buridiaditya another one, when without vision training occurs after the env restarts the client waits for 100 sec to restart. 
error

**Review by rudrasohan — APPROVED** [review]
- [x] Tested Single agent with vision.
- [x] Tested Single agent without vision.
- [x] Tested Multi agent without vision.

**Comment by rudrasohan** [comment]
There is one thing though as from a technical standpoint in both cases the without vision mode was working fine. But due to the frequent timeouts, I don't think the agent was able to learn anything. @buridiaditya is this a connection issue or that the simulator is literally insanely fast in the without viz mode.

File: MADRaS/envs/gym_madras.py [hunk 2]
```diff
@@ -13,30 +13,36 @@
...
+class MadrasEnv(TorcsEnv,gym.Env):
     """Definition of the Gym Madras Env."""
     def __init__(self, vision=False, throttle=True,
-                 gear_change=False, port=3001, pid_assist=True,
-                 CLIENT_MAX_STEPS=np.inf):
+                 gear_change=False, port=60934, pid_assist=False,
+                 CLIENT_MAX_STEPS=np.inf,visualise=True):
         """Init Method."""
+        self.torcs_proc = None
         self.pid_assist = pid_assist
```
Comment (MehaKaushik) [inline MADRaS/envs/gym_madras.py:37]: I would recommend writing a comment stating the use of the newly introduced variables.

File: MADRaS/envs/gym_madras.py [hunk 3]
```diff
@@ -75,13 +114,20 @@ def reset(self, prev_step_info=None):
 
         else:
             try:
-                self.ob, self.client =\
-                    TorcsEnv.reset(self, client=self.client, relaunch=True)
-
+                self.ob, self.client = TorcsEnv.reset(self, client=self.client, relaunch=True)
+                # command = None
+                # if self.visualise:
```
Comment (MehaKaushik) [inline MADRaS/envs/gym_madras.py]: Delete these commented lines.

File: MADRaS/utils/gym_torcs.py [hunk 4]
```diff
@@ -195,15 +198,32 @@ def reset(self, client, relaunch=False):
...
-        os.system('pkill torcs')
+    def reset_torcs(self,client):
+        print("relaunch torcs in gym_torcs on port{}".format(client.port))
+        
+        command = 'kill {}'.format(client.serverPID)
+        os.system(command)
+        
+        command = None
+        rank = MPI.COMM_WORLD.Get_rank()        
+        if rank == 0 and self.visualise:
```
Comment (MehaKaushik) [inline MADRaS/utils/gym_torcs.py]: As discussed offline, instead of rank == 0 , introduce a new variable, which could be set by the user.

**Review by MehaKaushik — COMMENTED** [review]
Good work here!! 

@buridiaditya I have left some minor comments, please address them.
@rudrasohan Please list down all the testings you did for this PR.

**Comment by buridiaditya** [comment]
@rudrasohan Finished testing with MPI. Approve the PR if nothing else is left.

**Pull request merged**